polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala

```diff
@@ -255,9 +271,12 @@ class PythonInterpreter(val kernelContext: KernelContext) extends LanguageInterp
             None
         }
 
-      case _ =>
+      case name =>
         Option(jep.getValue(accessor, classOf[PyObject])).map {
-          pyObj => new PythonObject(pyObj, runner) -> Some(typeOf[PythonObject])
+          pyObj =>
+            // make a refined PythonObject type with the name of the type as a constant string type
```

Comment:
This is where we construct (at compile time at runtime...) the literal constant type holding the python type name, and apply it to `TypedPythonObject`.